### PR TITLE
Feature/update to fog resolver

### PR DESCRIPTION
**Pull request by BrianCorbin**

### CODE REVIEW

File: full-service/src/bin/transaction-signer.rs [hunk 1]
```diff
@@ -352,7 +352,11 @@ fn tx_out_belongs_to_account(tx_out: &TxOut, account_view_private_key: &Ristrett
         Ok(k) => k,
     };
     let shared_secret = get_tx_out_shared_secret(account_view_private_key, &tx_out_public_key);
-    tx_out.masked_amount.get_value(&shared_secret).is_ok()
+    tx_out
+        .get_masked_amount()
+        .unwrap()
```
Comment (eranrund) [inline full-service/src/bin/transaction-signer.rs:357]: unwrap seems wrong here, it appears the function should return false if this fails.

File: full-service/src/config.rs [hunk 2]
```diff
@@ -120,7 +120,7 @@ impl APIConfig {
                 .build(),
         );
 
-        let conn = GrpcFogReportConnection::new(env, logger.clone());
+        let conn = GrpcFogReportConnection::new("".to_string(), env, logger.clone());
```
Comment (eranrund) [inline full-service/src/config.rs:123]: We should add a new command line argument `--chain-id` and pass it here.

File: full-service/src/config.rs [hunk 3]
```diff
@@ -213,6 +213,7 @@ impl PeersConfig {
             .iter()
             .map(|client_uri| {
                 ThickClient::new(
+                    "".to_string(),
```
Comment (eranrund) [inline full-service/src/config.rs:216]: We should add a new command line argument `--chain-id` and pass it here.

File: full-service/src/db/transaction_log.rs [hunk 4]
```diff
@@ -1069,7 +1069,7 @@ mod tests {
             &wallet_db.get_conn().unwrap(),
         )
         .unwrap();
-        assert_eq!(input_details0.value as u64, 7 * MOB);
+        assert_eq!(input_details0.value, associated_txos.inputs[0].value);
```
Comment (eranrund) [inline full-service/src/db/transaction_log.rs:1072]: What caused this change?

File: full-service/src/error.rs [hunk 5]
```diff
@@ -315,6 +315,9 @@ pub enum WalletTransactionBuilderError {
 
     /// Burn Redemption Memo must be exactly 128 characters (64 bytes) long.
     InvalidBurnRedemptionMemo(String),
+
+    /// Error converting a TxOut
+    TxOutConversion(mc_transaction_core::TxOutConversionError),
```
Comment (eranrund) [inline full-service/src/error.rs:320]: This error, as well as a few ones above it, fail to render the argument (they are missing the `: {0}` in the docstring).

File: full-service/src/json_rpc/v1/models/amount.rs [hunk 6]
```diff
@@ -60,14 +60,17 @@ impl TryFrom<&MaskedAmount> for mc_transaction_core::MaskedAmount {
             &hex::decode(&src.commitment)
                 .map_err(|err| format!("Could not decode hex for amount commitment: {:?}", err))?,
         );
-        Ok(Self {
+
+        let masked_amount = mc_transaction_core::MaskedAmountV1 {
```
Comment (eranrund) [inline full-service/src/json_rpc/v1/models/amount.rs:64]: This is scary... whats going to happen when we begin supporting masked amount v2? Maybe not in the v1 full-service API?

File: full-service/src/json_rpc/v1/models/receiver_receipt.rs [hunk 7]
```diff
@@ -95,7 +95,7 @@ mod tests {
         let account_key = AccountKey::random(&mut rng);
         let public_address = account_key.default_subaddress();
         let txo = TxOut::new(
-            BlockVersion::MAX,
+            BlockVersion::TWO,
```
Comment (eranrund) [inline full-service/src/json_rpc/v1/models/receiver_receipt.rs:98]: Why was this changed? Seems valuable to have a test that tests the newest block version.
In other places we put such tests in a loop that tries all relevant block versions.

File: full-service/src/json_rpc/v2/models/masked_amount.rs [hunk 8]
```diff
@@ -54,14 +54,17 @@ impl TryFrom<&MaskedAmount> for mc_transaction_core::MaskedAmount {
             &hex::decode(&src.commitment)
                 .map_err(|err| format!("Could not decode hex for amount commitment: {:?}", err))?,
         );
-        Ok(Self {
+
+        let masked_amount_v1 = mc_transaction_core::MaskedAmountV1 {
```
Comment (eranrund) [inline full-service/src/json_rpc/v2/models/masked_amount.rs:58]: The v2 API should plan to support MaskedAmountV2...

File: validator/service/src/validator_api.rs [hunk 9]
```diff
@@ -71,6 +71,7 @@ impl<UTC: UserTxConnection + 'static> ValidatorApi<UTC> {
             conn_manager,
             submit_node_offset: Arc::new(AtomicUsize::new(0)),
             fog_report_connection: GrpcFogReportConnection::new(
+                "".to_string(),
```
Comment (eranrund) [inline validator/service/src/validator_api.rs:74]: Need to be passed from command line

**Review by eranrund — CHANGES_REQUESTED** [review]
Spotted a few things that I think require attention:
1) tests that got pinned to an old block version (::TWO) - why not MAX? and can they instead be made to test all version (e.g. TWO..MAX)?
2) chain-id needs to be added to the command line config and used when creating grpc connections
3) MaskedAmountV2 is a thing that will need to be supported. Could be a future todo, but things will get messy once we switch to the block version that uses that.

File: full-service/src/json_rpc/v1/models/receiver_receipt.rs (hunk 7 above)
Comment (BrianCorbin) [reply full-service/src/json_rpc/v1/models/receiver_receipt.rs:98]: yeah agreed, was testing to see what would change. I'll put TODO comments from now on, bad habit of mine not to.

File: full-service/src/config.rs (hunk 3 above)
Comment (BrianCorbin) [reply full-service/src/config.rs:216]: ah okay, yeah that works

Comment (BrianCorbin) [reply full-service/src/config.rs:216]: Optional? Default value?

File: full-service/src/json_rpc/v2/models/masked_amount.rs (hunk 8 above)
Comment (BrianCorbin) [reply full-service/src/json_rpc/v2/models/masked_amount.rs:58]: agreed, will add that in 👍

File: full-service/src/config.rs (hunk 3 above)
Comment (eranrund) [reply full-service/src/config.rs:216]: Nope, required

Comment (BrianCorbin) [reply full-service/src/config.rs:216]: ah okay, thanks! Will add that as a required cla

Comment (BrianCorbin) [reply full-service/src/config.rs:216]: Where can I find info on chain-ids?

File: full-service/src/json_rpc/v1/models/receiver_receipt.rs (hunk 7 above)
Comment (BrianCorbin) [reply full-service/src/json_rpc/v1/models/receiver_receipt.rs:98]: Will create a gh issue to track that

File: full-service/src/db/transaction_log.rs (hunk 4 above)
Comment (BrianCorbin) [reply full-service/src/db/transaction_log.rs:1072]: This should no longer be relevant, that particular code in the test has changed to better reflect the desired outcome

File: full-service/src/bin/transaction-signer.rs (hunk 1 above)
Comment (BrianCorbin) [reply full-service/src/bin/transaction-signer.rs:357]: This code in particular will be addressed in the following PR #436

File: full-service/src/json_rpc/v1/models/receiver_receipt.rs (hunk 7 above)
Comment (eranrund) [reply full-service/src/json_rpc/v1/models/receiver_receipt.rs:98]: What breaks if it is set to ::MAX?

File: validator/service/src/config.rs [hunk 10]
```diff
@@ -24,4 +24,7 @@ pub struct Config {
     /// How many seconds to wait between polling.
     #[structopt(long, default_value = "5", parse(try_from_str=parse_duration_in_seconds))]
     pub poll_interval: Duration,
+
+    #[structopt(long)]
+    pub chain_id: String,
```
Comment (eranrund) [inline validator/service/src/config.rs:29]: Would become unnecessary once its moves into `PeersConfig`

File: full-service/src/config.rs (hunk 3 above)
Comment (eranrund) [reply full-service/src/config.rs]: https://github.com/mobilecoinfoundation/mcips/blob/main/text/0049-chain-ids.md

File: full-service/src/json_rpc/v1/models/receiver_receipt.rs (hunk 7 above)
Comment (BrianCorbin) [reply full-service/src/json_rpc/v1/models/receiver_receipt.rs:98]: Nothing now, it should be back to MAX

Comment (BrianCorbin) [reply full-service/src/json_rpc/v1/models/receiver_receipt.rs:98]: oh wait I missed this one, hahha

File: full-service/src/service/receipt.rs [hunk 11]
```diff
@@ -144,7 +156,16 @@ impl TryFrom<&mc_api::external::Receipt> for ReceiverReceipt {
         let public_key: CompressedRistrettoPublic =
             CompressedRistrettoPublic::try_from(src.get_public_key())?;
         let confirmation = TxOutConfirmationNumber::try_from(src.get_confirmation())?;
-        let amount = MaskedAmount::try_from(src.get_masked_amount())?;
+
+        let amount = if let Ok(masked_amount_v2) = MaskedAmountV2::try_from(src.get_masked_amount())
```
[the remark this answers is not in the document]
Comment (BrianCorbin) [reply full-service/src/service/receipt.rs:160]: To try to decode to V2 first, and then V1, which seems to be working

[the remark this answers is not in the document]
Comment (eranrund) [reply full-service/src/service/receipt.rs:160]: Isn't this just a coincidence based on what block version the tests are using? If the fields are identical when would MaskedAmountV2::try_from ever fail and MaskedAmountV1::try_from succeed?

[the remark this answers is not in the document]
Comment (eranrund) [reply full-service/src/service/receipt.rs:160]: It looks like @garbageslam put out https://github.com/mobilecoinfoundation/mobilecoin/pull/2532 so maybe upreving to that once it gets merged in a bit will remove the confusion here?

[the remark this answers is not in the document]
Comment (BrianCorbin) [reply full-service/src/service/receipt.rs:160]: Yeah that's a good idea, let's wait on that

File: full-service/src/json_rpc/v1/models/amount.rs [hunk 12]
```diff
@@ -27,26 +39,24 @@ pub struct MaskedAmount {
...
-    fn from(src: &mc_api::external::MaskedAmount) -> Self {
-        Self {
-            object: "amount".to_string(),
-            commitment: hex::encode(src.get_commitment().get_data()),
-            masked_value: src.get_masked_value().to_string(),
-            masked_token_id: hex::encode(&src.get_masked_token_id()),
-        }
-    }
+    /// The version of the masked amount.
+    pub version: MaskedAmountVersion,
```
[the remark this answers is not in the document]
Comment (BrianCorbin) [reply full-service/src/json_rpc/v1/models/amount.rs:44]: I'm tempted to just remove the receipt stuff from v1 entirely and make it v2 only at this point like I did with view only accounts.

[the remark this answers is not in the document]
Comment (BrianCorbin) [reply full-service/src/json_rpc/v1/models/amount.rs:44]: but also it really doesn't matter that much

[the remark this answers is not in the document]
Comment (BrianCorbin) [reply full-service/src/json_rpc/v1/models/amount.rs:44]: no, only MOB

[the remark this answers is not in the document]
Comment (eranrund) [reply full-service/src/json_rpc/v1/models/amount.rs:44]: then I think it's safe for the V1 API to only support MaskedAmountV1

[the remark this answers is not in the document]
Comment (eranrund) [reply full-service/src/json_rpc/v1/models/amount.rs:43]: What made you resolve this? I think this still breaks the v1 API since there's now a new required field that didn't exist before

[the remark this answers is not in the document]
Comment (BrianCorbin) [reply full-service/src/json_rpc/v1/models/amount.rs:43]: Ah, I misunderstood, this should always assume MaskedAmountV1

**Review by eranrund — APPROVED** [review]
